src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/PoolOfThings.cs

```diff
@@ -0,0 +1,374 @@
+        }
+
+        public void Dispose()
+        {
+            DestoryAllPooledInstances();
+        }
+
+        // Replaces the pool of channels with a new one and destory all instances in the previous pool
+        public void DestoryAllPooledInstances()
+        {
+            // Get the new pool ready
+            T[] newPool = new T[_maxSize];
+
+            // Replace the pool
+            var oldPool = _thePool;
+            if (Interlocked.CompareExchange(ref _thePool, newPool, oldPool) != oldPool)
+            {
+                // somebody beat us - they will be responsible for closing the old _channelsPool they replaced
+            }
+            else
+            {
+                // we are the one who replaced the old _thePool - we're responsible for cleaning it up
+                DestroyAllPooledInstanciesImpl(oldPool, _destroyInstance);
+            }
+        }
+
+        public IEnumerable<T> GetAllPooledInstances()
+        {
+            return new AllPooledInstancesCollection(this);
+        }
+
+        #region helpers
+
+        private class AllPooledInstancesCollection : IEnumerable<T>
+        {
+            private PoolOfThings<T> _thePool;
+            public AllPooledInstancesCollection(PoolOfThings<T> thePool)
+            {
+                _thePool = thePool;
+            }
+
+            public IEnumerator<T> GetEnumerator()
+            {
+                for (int i = 0; i < _thePool._thePool.Length; i++)
+                {
+                    var instance = _thePool.GetPooledInstance(i);
+                    // skip nulls 
+                    if (instance != default(T))
+                    {
+                        yield return instance;
+                    }
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+        }
+
+        private T GetPooledInstance(int i)
+        {
+            var instance = _thePool[i];
+            while (instance == null)
+            {
+                //try
+                {
+                    instance = _createInstance();
+                    if (instance == null)
+                    {
+                        break;
+                    }
+                    if (Interlocked.CompareExchange<T>(ref _thePool[i], instance, null) != null)
+                    {
+                        _destroyInstance(instance);
+                        instance = _thePool[i];
+                        // Somebody stored the new instance before us - dispose ours and re-read theirs
+                        // Since the whole pool can also be replaced we can still get a null from _thePool[index]
+                        // so we retry while (instance == null)
+                        // Note that _destroyInstance(instance) should not throw since nobody uses it yet
+
+                        //Interlocked.Increment(ref _numPoolSetRetries);
+                    }
```

Comment:
Not for this code review but future -- can that commented out Interlocked be removed?  Or is it something that should be uncommented?  One style rule for OSS is to avoid "dead code" in comments.